File: cpp/programmers/프린터.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

int solution(vector<int> priorities, int location) {
    int answer = 0, cnt = 0;
    priority_queue<int> pq;
    queue<pair<int, int>> que;
    
    for (int i=0; i<priorities.size(); ++i){
        que.push({priorities[i], i});
        pq.push(priorities[i]);
    }
    
    while (true){
        int priority = que.front().first;
        int position = que.front().second;
        int max_pri = pq.top();
        que.pop();
        if (priority == max_pri){
            ++cnt;
            if (position == location){
                answer = cnt;
                break;
            }
            else{
                pq.pop();
            }
        }
        else{
            que.push({priority, position});
        }
        
    }
    
    return answer;
}

File: python/programmers/프린터.py
from collections import deque


def solution(priorities, location):
    answer = 0
    queue = deque()
    for i in range(len(priorities)):
        queue.append([priorities[i], i])
    cnt = 0
    while True:
        max_pri = 0
        for i in range(len(queue)):
            max_pri = max(queue[i][0], max_pri)
        priority, position = queue.popleft()
        if priority == max_pri:
            cnt += 1
            if position == location:
                answer = cnt
                break
        if priority < max_pri:
            queue.append([priority, position])

    return answer